### PR TITLE
Meson test fixups

### PR DESCRIPTION
**Pull request by eli-schwartz**
- don't make running the valgrind tests mandatory
- mark known-broken test on Windows, to work around #3119

### CODE REVIEW

**Comment by eli-schwartz** [comment]
After manually installing these fixes into the Meson WrapDB, I finally got both Windows and macOS (new additions to our test matrix) to pass CI: https://github.com/mesonbuild/wrapdb/runs/6089540648?check_suite_focus=true

File: build/meson/tests/meson.build [hunk 1]
```diff
@@ -183,6 +185,8 @@ test('test-zstream-1',
 test('test-zstream-3',
   zstreamtest,
   args: ['--newapi', '-t1', ZSTREAM_TESTTIME] + FUZZER_FLAGS,
+  # --newapi dies on Windows with "exit status 3221225477 or signal 3221225349 SIGinvalid"
+  should_fail: host_machine_os == os_windows,
```
Comment (Cyan4973) [inline build/meson/tests/meson.build:196]: what if the issue is fixed, and `zstreamtest --newapi` now completes successfully ?
would it be considered a test failure due to `should_fail` ?

Comment (eli-schwartz) [reply build/meson/tests/meson.build:196]: This PR changes its status to EXPECTEDFAIL, which is a success status. When the bug is fixed, it will start producing an UNEXPECTEDPASS status, which is considered a failure and causes the overall test results to be negative.

Comment (Cyan4973) [reply build/meson/tests/meson.build:196]: > This PR changes its status to EXPECTEDFAIL, which is a success status. When the bug is fixed, it will start producing an UNEXPECTEDPASS status, which is considered a failure and causes the overall test results to be negative.

Better fix the issue then

Comment (eli-schwartz) [reply build/meson/tests/meson.build:196]: Yep, once that other PR is merged I will re-check this and drop the obsolete commit.

**Comment by eli-schwartz** [comment]
Is there anything else I should do about this?

As far as discussion over the second commit goes, it seems the attempt to actually fix the issue isn't getting anywhere. I think it's fine to add a "known broken" label to the testsuite.

**Comment by Cyan4973** [comment]
I don't remember all the details,
it seems that : 
- On Windows + MinGW, and likely Windows + Visual, `zstreamtest --newpi` fails sometimes. But it's not reliably reproducible, meaning it fails at different places between consecutive runs on the same exact test.
- This seems to points at a race condition in the multi-threading code. And it's possibly related to the shim `pthread` translation layer for Windows, since there is no such issue when using `pthread` directly, as in Windows + Cygwin combination for example, which works fine.
- The proposal in this PR is to hide the issue by disabling this test, or making its result non-blocking, when run on Windows.

I think it's fair to say that I would largely prefer to have the issue fixed.
Problem is, the fix is not yet found, and it's a hard problem because the issue is very difficult to observe.
And a previous fix attempt, in #3121, failed to reach the objective.

Given a choice, I would still prefer to fix the issue.
Problem is, finding the time to look into it is difficult, our team is tied by other priorities.
We can still make this item a priority objective for the next release, but it doesn't tell when someone will take care of it.
It just imply that it's probably going to happen this year.

**Comment by eli-schwartz** [comment]
> The proposal in this PR is to hide the issue by disabling this test, or making its result non-blocking, when run on Windows.

To be clear, the proposal is to make it be non-blocking, but visibly reported:

```
passing tests, yay : XXX
failing tests, phew: 0
known broken? :/   : 1
```

This is something that the Test Anything protocol calls "todo" tests: https://testanything.org/tap-specification.html#todo-tests
> These tests represent a feature to be implemented or a bug to be fixed and act as something of an executable “things to do” list.

Automake's ["Generalities about testing"](https://www.gnu.org/software/automake/manual/html_node/Generalities-about-Testing.html) mentions:
> It’s not uncommon, especially during early development stages, that some tests fail for known reasons, and that the developer doesn’t want to tackle these failures immediately (this is especially true when the failing tests deal with corner cases).

It's not uncommon to make a test failure be non-blocking when you've temporarily exhausted your efforts at solving it (and you have a tracking issue for it, *and* the testsuite flags it as a FIXME). If the test failure doesn't indicate a critical release blocker, then you want the testsuite status to be useful -- i.e. it should be non-blocking by default, and only raise critical errors if new or unknown issues are introduced.

**Comment by Cyan4973** [comment]
OK, this will be discussed with the team, so that we can provide a feedback on this PR in a reasonable delay.

**Comment by eli-schwartz** [comment]
Alright, thanks for looking into it. :)

**Comment by Cyan4973** [comment]
Follow up : @terrelln is going to have a look at this topic.
It requires a working Windows machine, with development tools, so it's gonna take a little time, but hopefully nothing too long.
The hope is that a fix can be found quickly, on the ground that the problem can be reproduced pretty easily.
But if not, we'll revisit this proposal of having a non-blocking test while waiting for a (longer term) fix.

**Comment by eli-schwartz** [comment]
Any update on this? :)

**Comment by eli-schwartz** [comment]
@terrelln did you ever have a chance to investigate this?

**Comment by yoniko** [comment]
> @terrelln did you ever have a chance to investigate this?

I'll take a look at this today / tomorrow.

**Comment by yoniko** [comment]
An update: I've reproduced the problem and found the root cause.
There's indeed a bug in the windows threading translation layer.
Will put up a PR with a fix soon.

**Comment by eli-schwartz** [comment]
Fantastic! Thank you for your investigation. Looking forward to being able to pass tests on the next zstd release. :)

**Comment by eli-schwartz** [comment]
Added CI improvements on top. This also includes #3368 which is needed to configure the contrib directory so that CI runs, although that PR is much more easily merged.

File: .github/workflows/dev-short-tests.yml [hunk 2]
```diff
@@ -178,6 +178,51 @@ jobs:
...
+          meson install -C builddir --destdir staging/
+
+  meson-windows:
+    runs-on: windows-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Install packages
+        run: pip install --pre meson
+      - name: Initialize the MSVC dev command prompt
+        uses: ilammy/msvc-dev-cmd@v1
```
Comment (yoniko) [inline .github/workflows/dev-short-tests.yml]: We should specify a commit hash rather than a tag, especially when depending on third-party actions.

**Review by yoniko — CHANGES_REQUESTED** [review]
Looks good, my only feedback is that I think it'd be better to pin the third-party action to a specific commit.
After making the change I think it'd be a good idea to merge this and then merge a PR that fixes the issue and removes the expected failure for windows.

File: .github/workflows/dev-short-tests.yml (hunk 2 above)
Comment (eli-schwartz) [reply .github/workflows/dev-short-tests.yml]: Fair enough. :) I tend to avoid third-party ones too, at least for workflows that have secrets! but I don't usually bother with hash pinning since I either trust the action source or don't give it secrets -- and I copied this from a workflow that doesn't have secrets.

I'll make this change when rebasing (soon).

Comment (yoniko) [reply .github/workflows/dev-short-tests.yml]: Yeah, it's probably safe, but at the same time it's a better practice to pin to specific hashes and it's a change we plan on doing with other dependencies as well.
In any case, sounds good.

Comment (eli-schwartz) [reply .github/workflows/dev-short-tests.yml]: Done.

**Comment by yoniko** [comment]
Looks good, although a bit unstable as the windows test might actually pass sometimes.
Merging so we can merge the other PR that solves the bug and would fix the CI to expect to pass.

**Comment by eli-schwartz** [comment]
Oops, yeah, I think I grabbed a hash for the msvc-dev-cmd action, from the wrong branch.

**Comment by eli-schwartz** [comment]
Honestly, I really wish it was possible to target semantically meaningful versions of an action, like I originally did, and specify a checksum to verify, instead of abusing commit hashes. But GitHub doesn't support that. :(
